sdk/servicebus/Microsoft.Azure.ServiceBus/src/Primitives/ConcurrentExpiringSet.cs

```diff
@@ -4,64 +4,71 @@
 namespace Microsoft.Azure.ServiceBus.Primitives
 {
     using System;
+    using System.Collections.Generic;
     using System.Collections.Concurrent;
+    using System.Threading;
     using System.Threading.Tasks;
 
     sealed class ConcurrentExpiringSet<TKey>
     {
         readonly ConcurrentDictionary<TKey, DateTime> dictionary;
-        readonly object cleanupSynObject = new object();
-        bool cleanupScheduled;
-        static TimeSpan delayBetweenCleanups = TimeSpan.FromSeconds(30);
+        readonly ICollection<KeyValuePair<TKey, DateTime>> dictionaryAsCollection;
+        CancellationTokenSource tokenSource = new CancellationTokenSource(); // doesn't need to be disposed because it doesn't own a timer
+        volatile TaskCompletionSource<bool> cleanupTaskCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+        static readonly TimeSpan delayBetweenCleanups = TimeSpan.FromSeconds(30);
 
         public ConcurrentExpiringSet()
         {
             this.dictionary = new ConcurrentDictionary<TKey, DateTime>();
+            this.dictionaryAsCollection = dictionary;
+            _ = CollectExpiredEntriesAsync(tokenSource.Token);
         }
 
         public void AddOrUpdate(TKey key, DateTime expiration)
         {
             this.dictionary[key] = expiration;
-            this.ScheduleCleanup();
+            this.cleanupTaskCompletionSource.TrySetResult(true);
         }
 
         public bool Contains(TKey key)
         {
             return this.dictionary.TryGetValue(key, out var expiration) && expiration > DateTime.UtcNow;
         }
 
-        void ScheduleCleanup()
+        public void Clear()
         {
-            lock (this.cleanupSynObject)
-            {
-                if (this.cleanupScheduled || this.dictionary.Count <= 0)
-                {
-                    return;
-                }
-
-                this.cleanupScheduled = true;
-                Task.Run(async () => await this.CollectExpiredEntriesAsync().ConfigureAwait(false));
-            }
+            this.tokenSource.Cancel();
+            this.dictionary.Clear();
+            this.cleanupTaskCompletionSource.TrySetCanceled();
+            this.cleanupTaskCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            this.tokenSource = new CancellationTokenSource();
+            _ = CollectExpiredEntriesAsync(tokenSource.Token);
```

Comment:
>CollectExpiredEntriesAsync [](start = 16, length = 26)

When will the final task be cancelled? Looks like its going to be running forever at `await cleanupTaskCompletionSource.Task.ConfigureAwait(false);`

Reply:
See related change in MessageReceiver

Reply:
That just calls `Clear()`. But Inside this method you are calling new async Task which is never cancelled.

Reply:
Good catch. Not sure why I missed that. Will fix